### PR TITLE
Table of contents

### PR DESCRIPTION
**Pull request by thehenster**
(I have slowly added commits to this and hence have edited this description down in case the email notifications are confusing)

This is now a gem that is referenced locally from the Gemfile. There's no need to release it yet, but if we stick with the "let them fork it" approach then moving shared functionality into gems is one way of avoiding them having to merge in upstream changes.

### CODE REVIEW

File: Gemfile.lock [hunk 1]
```diff
@@ -1,13 +1,13 @@
 GEM
   remote: https://rubygems.org/
   specs:
-    activesupport (4.2.7.1)
+    activesupport (5.0.0.1)
```
Comment (thehenster) [inline Gemfile.lock]: This got bumped to 5 when resolving a rebase conflict and I regenerated the Gemfile.lock. If this wasn't still in development I'd avoid this but as it's not.. It's not harmful.

**Comment by thehenster** [comment]
It would be nice to remove the nokogiri dependency if possible. Nokogiri is fast and forgiving with malformed HTML, but we don't really care about speed and the HTML is generated from the markdown. It might be possible to replace it with REXML.

**Comment by thehenster** [comment]
I think I might remove the keyword arguments to make this Ruby 1.9 compatible. It looks like middleman is 1.9 compatible and it won't be "our" ruby interpreter that runs this thing. Hmm.

It's mergable though, bar any other feedback. Improvements can always come later.

**Comment by 36degrees** [comment]
So I have recently learnt to read, and guess what I found.

> The HTML renderer has an alternate version, Redcarpet::Render::HTML_TOC, which will output a table of contents in HTML based on the headers of the Markdown document.
> 
> When instantiating this render object, you can optionally pass a nesting_level option which takes an integer and allows you to make it render only headers until a specific level.
> 
> - https://github.com/vmg/redcarpet

Le facepalm.

**Comment by thehenster** [comment]
I have finally remembered why I didn't use it! I'm now wondering if it's not a very good reason..

That `Redcarpet::Render::HTML_TOC` renderer wants markdown, not HTML, and `yield` returns the compiled HTML. I thought at the time that fighting the framework to return markdown rather than HTML wasn't worth it. I'll try fighting it for a bit to see how plausible it is. It probably won't be that hard actually :/

**Comment by 36degrees** [comment]
If it's a lot of work, happy with your solution – the pre-existing one just seems slightly less fragile as it's not trying to parse HTML. But that sounds like it might be a valid reason not to use it.

**Comment by thehenster** [comment]
I had a play. It might be possible to use `Redcarpet::Render::HTML_TOC` but not until we have time to do a deeper dive I don't think either of the ways I thought of are very nice: https://github.com/alphagov/tech-docs-template/pull/19

I'm not too worried about the fragility of the HTML. It's just been generated from markdown so it should be well formed.

It's a little sad that this adds Nokogiri as a dependency.. but until we've got time for a deeper dive I think we should go with this one. It's vendored and so will be easy to delete/replace when/if the time comes. It also doesn't affect the way you use a normal middleman app, apart from the `#table_of_contents` method itself.

File: vendor/table_of_contents/lib/table_of_contents/heading.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
+module TableOfContents
+  class Heading
+    def initialize(element_name:, text:, attributes:)
+      @element_name = element_name
+      @text = text
+      @attributes = attributes
+    end
+
+    def size
+      @element_name.scan(/h(\d)/) && $1 && Integer($1)
```
Comment (mcmillan) [inline vendor/table_of_contents/lib/table_of_contents/heading.rb]: This is more of a personal opinion than anything else, but I prefer avoiding `$1` et al and doing stuff to the `MatchData` class that's returned from `String.match`; something akin to

```ruby
@element_name.match(/h(\d)/)&.captures&.first&.to_i
``` 

Actually I don't know if that's any better. Meh.

Comment (thehenster) [inline vendor/table_of_contents/lib/table_of_contents/heading.rb]: It's a similar devil..

Comment (mcmillan) [inline vendor/table_of_contents/lib/table_of_contents/heading.rb]: Yeah I actually think the above is even worse. Ignore me.

**Pull request merged**